File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Java/src/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author guillermo
 */
public class Napakalaki {
    //Atributos
    final private static Napakalaki instance = new Napakalaki();
    private Monster currentMonster;
    private CardDealer dealer;
    private ArrayList<Player> players;
    private Player currentPlayer;
    
    //Constructor
    private Napakalaki(){
        players = new ArrayList();
        dealer = CardDealer.getInstance();
    }
    
    //Métodos
    private void initPlayers(ArrayList<String> names){
        for(String name: names)
            players.add(new Player(name));
    }
    
    private Player nextPlayer(){
        Random rand = new Random();
        if(currentPlayer == null)
            currentPlayer = players.get(rand.nextInt(players.size()));
        else 
            currentPlayer = players.get(
                    (players.indexOf(currentPlayer)+1) % players.size()
            );
        return currentPlayer;
    }
    
    private boolean nextTurnIsAllowed(){
        if(currentPlayer == null)
            return true;
        else
            return currentPlayer.validState();
    }
    
    private void setEnemies(){
        Random rand = new Random();
        int i = rand.nextInt(players.size()-1)+1;
        for(Player p: players){
            p.setEnemy(players.get(i));
            i = (++i)%players.size();
        }
    }
    
    static public Napakalaki getInstance(){
        return instance;
    }
    
    public CombatResult developCombat(){
        CombatResult cr = currentPlayer.combat(currentMonster);
        if (cr == CombatResult.LOSEANDCONVERT) {
            CultistPlayer nuevo = new CultistPlayer(currentPlayer,dealer.nextCultist());
            for(Player p:players){
                if(p.getEnemy() == currentPlayer)
                    p.enemy = currentPlayer;
            }
            players.remove(currentPlayer);
            currentPlayer = nuevo;
            players.add(currentPlayer);
        }
        dealer.giveMonsterBack(currentMonster);
        return cr;        
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        for(Treasure t: treasures){
            currentPlayer.discardVisibleTreasure(t);
            dealer.giveTreasureBack(t);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        for(Treasure t: treasures){
            currentPlayer.discardHiddenTreasure(t);
            dealer.giveTreasureBack(t);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for(Treasure t: treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame(ArrayList<String> players){
        initPlayers(players);
        setEnemies();
        dealer.initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        //TODO
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOk = nextTurnIsAllowed();
        if(stateOk){
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
            boolean dead = currentPlayer.isDead();
            if(dead)
                currentPlayer.initTreasures();
        }
        return stateOk;
    }
    
    public boolean endOfGame(CombatResult result){
        return result == CombatResult.WINGAME;
    }
}

File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Ruby/lib/napakalaki.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require 'singleton.rb'
require_relative 'card_dealer.rb'
require_relative 'combat_result.rb'
require_relative 'monster.rb'
require_relative 'player.rb'
require_relative 'cultist_player.rb'

class Napakalaki
  include Singleton
  attr_reader :currentPlayer, :currentMonster
  
  def initialize
    @currentPlayer = nil
    @players = Array.new
    @dealer = CardDealer.instance
    @currentMonster = nil
  end
  
  def getCurrentPlayer
    @currentPlayer
  end
  
  def getCurrentMonster
    @currentMonster
  end
  
  def initPlayers(names)
    for n in names do
      jugador = Player.new(n)
      @players << jugador
    end
  end
  
  def nextPlayer
    if(@currentPlayer == nil ) then
      @currentPlayer = @players[rand(@players.size)]
    else
      @currentPlayer = @players[(@players.index(@currentPlayer)+1) % @players.size]
    end
    @currentPlayer
  end
  
  def nextTurnIsAllowed
    if(@currentPlayer == nil)
      return true
    else
      return @currentPlayer.validState
    end
  end
  
  def setEnemies
    i = rand(@players.size - 1) + 1
    for p in @players do
      p.enemy = @players[i]
      i = (i+1) % @players.size
    end
  end
  
  def developCombat
    cr = @currentPlayer.combat(@currentMonster)
    @dealer.giveMonsterBack(@currentMonster)
    if(cr == CombatResult::LOSEANDCONVERT)
      cultist = @dealer.nextCultist
      cp = CultistPlayer.new(@currentPlayer, cultist)
      
      #Cambiar de lista de jugadores
      i = @players.index(@currentPlayer)
      @players.delete_at(i)
      @players.insert(i,cp)
      
      #Cambiar de enemigos
      for p in @players do
        if (p.enemy == @currentPlayer)
          p.enemy = cp
        end
      end
      
      #Cambiar de currentPlayer
      @currentPlayer = cp
      
    end
    return cr
  end
  
  def discardVisibleTreasures(treasures)
    for t in treasures do
      @currentPlayer.discardVisibleTreasure(t)
      @dealer.giveTreasureBack(t)
    end
  end
  
  def discardHiddenTreasures(treasures)
    for t in treasures do
      @currentPlayer.discardHiddenTreasure(t)
      @dealer.giveTreasureBack(t)
    end
  end
  
  def makeTreasuresVisible(treasures)
    for t in treasures do
      @currentPlayer.makeTreasureVisible(t);
    end
  end
  
  def initGame(players)
    initPlayers(players)
    setEnemies
    @dealer.initCards
    nextTurn
  end
  
  def nextTurn
    stateOk = nextTurnIsAllowed
    if(stateOk) then
      @currentMonster = @dealer.nextMonster
      @currentPlayer = nextPlayer
      dead = @currentPlayer.isDead
      if(dead) then
        @currentPlayer.initTreasures
      end
    end
    stateOk
  end
  
  def endOfGame(result)
    result == CombatResult::WINGAME;
  end
  
  private :initPlayers, :nextPlayer, :nextTurnIsAllowed, :setEnemies
end
